### ISSUE TITLE
ENH: Add a np.isnan loop for the object dtype

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#9009]
This would fix help fix #9008.

From most to least-well defined
 1. Object arrays of `float`,`np.float32`, `np.datetime64`... &rarr; `return np.isnan(item)`
 2. Object arrays of `int`,`np.int32`, ... &rarr; `return False`
 3. Object arrays of any type that implements `__float__` or `__int__` - return `np.isnan(float(x))`
 4. Object arrays of `np.ndarray` &rarr; `return np.isnan(item)` (requires a `O->O` loop, not `O->?`)

I think I'd draw the line after 3, and have 4 throw an error.

**Comment by h-vetinari** [comment]
I'd draw the line at 4. `nan` is a ubiquitous missing-value marker for many downstream libraries (not least pandas), and that may just as well happen in an array of objects. Not 100% sure I understand "Object arrays of `np.ndarray`", assuming you don't mean nested arrays.

I think it would be very reasonable to have:
```
arr = np.array([1, '2', {3, '3'}, np.nan])
arr
# array([1, '2', {3, '3'}, nan], dtype=object)
np.isnan(arr)  # <- currently raises
# array([False, False, False,  True])
```

**Comment by eric-wieser** [comment]
> assuming you don't mean nested arrays.

No, that _is_ what I mean

**Comment by h-vetinari** [comment]
> No, that is what I mean

To be fair, it wasn't clear whether the ellipses in 1. & 2. would include things such as strings, lists, maps, etc. But in any case, then I guess 3 is a good cut-off.

**Comment by smithsp** [comment]
Would this enhancement allow 
```python
numpy.isnan(numpy.array([],dtype=object)))
```
to not fail?

**Comment by mattip** [comment]
@eric-wieser is this still relevant?

**Comment by eric-wieser** [comment]
Yes, I think so. I think #6320 contains a candidate implementation.

**Comment by h-vetinari** [comment]
People are [bending](https://stackoverflow.com/questions/36198118/np-isnan-on-arrays-of-dtype-object) themselves out of shape to do this. `isnan` and friends simply shouldn't fail on object, the error message is detrimental IMO:
```
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```
because testing for nan's (at least in principle) doesn't have anything to do with having to cast.
